### DISCUSSION TITLE
How to show a single Volume Pixel by vol.legosurface?

### DISCUSSION THREAD

**Discussion by CCHLENG** [marcomusy/vedo#809] [Q&A] [answered]
[this example](https://github.com/marcomusy/vedo/blob/master/examples/volumetric/numpy2volume2.py) code can show this:
![image](https://user-images.githubusercontent.com/25116183/218710201-d0f803d8-5af1-4c36-9666-1864de6590b9.png)

I add
```python
data_matrix[1, 1, 40] = 3
```
and then 
```python
show([vol, lego], N=2, axes=1).close()
```
I get this:
![image](https://user-images.githubusercontent.com/25116183/218710921-5b009d99-6518-486c-8aec-2f6820fefede.png)

point `data_matrix[1, 1, 40]` can be seen in left side but any Volume Pixel in right side.

if I use `invert=True` in vol.legosurface, this fig is shown. Point `data_matrix[1, 1, 40]` is missed if inverted. why this volume pixel can't be shown when `invert=False`?
![image](https://user-images.githubusercontent.com/25116183/218711560-c87cd68f-6745-45e9-9607-92228fd45ae2.png)

how to show a single Volume Pixel by vol.legosurface? thanks!

**Comment by marcomusy** [comment]
Hi, this is a bit tricky but the value from the numpy array are not associated to the voxel center but to the vertices.
Note that in your last image you see 4 little cubes missing (not just one voxel or cell).
This normally is irrelevant as one deals with large dataset, but may become relevant for very small datasets. 
See eg:

```python

from vedo import *

data_matrix = np.zeros([70, 80, 90], dtype=float)
data_matrix[0:30,   0:30,  0:30] = 1.
data_matrix[30:50, 30:60, 30:70] = 2.
data_matrix[50:70, 60:80, 70:90] = 3.
data_matrix[1, 1, 40] = 3.

vol = Volume(data_matrix, c=['white','b','g','r'], mode=1)

print(vol.pointdata.keys())
print(vol.celldata.keys())

pts = Points([[0,0,0],[30,30,30], [50,60,70], [50,60,70]]).ps(10)

lego = vol.legosurface(1.5,3.5, invert=0, boundary=1)

show(vol.box(), lego, pts, axes=1).close()
```
![Screenshot from 2023-02-14 15-58-48](https://user-images.githubusercontent.com/32848391/218776571-042b37fa-b544-426a-82da-44b596b645eb.png)

**Comment by CCHLENG** [comment, marked as answer]
Thanks for your reply!

In your demo, `boundary=1` in `vol.legosurface` make the little cube visible.
Enlighten by numpy array are associated to the vertices, I can use 
```python
data_matrix[0, 0, 0] = 3
 vol = Volume(data_matrix, mode=1, origin=[-0.5, -0.5, -0.5])
``` 
to show a single cube like this:
![image](https://user-images.githubusercontent.com/25116183/218902386-f6c3a25c-723a-4022-a6a4-a6c405aa37b2.png)

But I observed that the color of legosurface is error when I add `lego.cmap('hot_r', vmin=1, vmax=3.5).add_scalarbar3d()` after `lego = vol.legosurface(1.0, 3.5, boundary=1)` .The middle large dataset and the single cube at [0,0,0] should be yellow. What can I do to show the right color?
![image](https://user-images.githubusercontent.com/25116183/218903681-6baf769c-6692-4357-84aa-d5878f651fa7.png)

Actually, what I want to visualize is a 3D occupancy grid map which created from a points cloud. 
For example, if there is any points in cube(x[-0.5,0.5],y[-0.5,0.5],z[-0.5,0.5]), the cube which center is [0,0,0] should be shown. 
The final result may like [this](https://github.com/marcomusy/vedo/blob/master/examples/volumetric/legosurface.py), many cube flock together, and some cube is single.
I realized that I also can draw it by `box` , but it might be hard to calculate every center and color of the box. Is there a easy way to show a 3D occupancy grid map?

**Reply by marcomusy** [reply]
Hi what about this:
`examples/pyplot/histo_3d.py`

![Screenshot from 2023-02-16 19-17-15](https://user-images.githubusercontent.com/32848391/219453114-d8d4f726-1e99-48d1-99be-9e7b5f2d48a7.png)

**Reply by CCHLENG** [reply]
It's like what I want! Thanks!
The link in [vedo gallery ](https://vedo.embl.es/#gallery) need to be fix, I have found that figure but the link is error. The 'D' of 'histo_3D.py' in the link should be lowercase.
![image](https://user-images.githubusercontent.com/25116183/219527675-856d5840-4c81-4ad3-9b65-06a3d5b05573.png)

In `histo_3D.py` ,
```python
pcloud = Volume(counts).topoints()
marker = Cube()
glyphed_pcl = Glyph(pcloud, marker, color_by_scalar=True)
```
`Glyph` creates marker at every points of pcloud.
What's different between marker is `cube` and `box`? I found that when I set marker as  `box` ,the surface is teared, but `cube` works as expect.
Ang how can I create `pcloud` directly? from my pointclould and value? In the demo,  `pcloud` is created by `.topoints()` method, which return a vedo Points.
```python
    ...
    def topoints(self):
        ...
        mpts = vedo.Points(v2p.GetOutput())
        return mpts
    ...
```
Which method should I use to set the value of every point?

**Reply by marcomusy** [reply]
Hi,
thanks I will fix the link!
You can create a pointcloud directly by passing 2D/3D coordinates, e.g.
```python
from vedo import *

pts = np.random.rand(100,3)
pcloud = Points(pts)

# add any data to each point
pcloud.pointdata["myarray1"] = np.sin(pts[:,0])
# pcloud.pointdata["myarray2"] = np.sin(pts[:,1])

marker = Cube().scale(0.1)
# pcloud.pointdata.select("myarray1")
glyph = Glyph(pcloud, marker, scale_by_scalar=True)
glyph.cmap('rainbow').add_scalarbar('counts')

show(glyph, axes=1)
```

<img width="383" alt="Screenshot 2023-02-18 at 14 26 20" src="https://user-images.githubusercontent.com/32848391/219868616-21f8e0bf-a830-401d-ba9c-4d7ab339f8dc.png">

**Reply by CCHLENG** [reply]
It works! Thanks! 
Do you have a donate link?

**Reply by marcomusy** [reply]
> Do you have a donate link?

I will set up one! :)

Let me know if need more help.
